### DISCUSSION TITLE
Alternative Interfaces to Step/Dir (CAN/RS485/Etc.)

### DISCUSSION THREAD

**Discussion by graykimchi** [grblHAL/core#524] [Q&A] [answered]
Hi,

Im just starting to look into grblHAL! This is an amazing project and I look forward to learning so much through it! (thank you everyone!)

I was wondering if there was any provision or pointers I could get to integrate a serial protocol like CAN (CANopen Software Layer), RS485, Etc. to control servo drives instead of STEP/DIR. 

I am looking to control closed-loop LV BLDC and HV PMSM servo drives. 

I see that the ISRs for generating step pulses are tied to a bitfield inside - [stepdir_map.h](http://svn.io-engineering.com/grblHAL/html/stepdir__map_8h.html) - that updates Step/dirs registers atomically so it looks like its pretty baked into the grblHAL code base but I was wondering if anyone has done such development and would be willing to teach me the ways

Thanks,
Gray

**Comment by terjeio** [comment, marked as answer]
As a start you can hook your (plugin?) code into [hal.stepper.pulse_start](http://svn.io-engineering.com/grblHAL/html/hal_8h.html#ab02e12b19a0b0015f5040340acbb7839) and send commands to the stepper drivers when a new segment arrives. A new segment is started every ~5ms and the acceleration and feed rate is constant during each.
FYI spindle sync is implemented by swapping `hal.stepper.pulse_start` between [normal](https://github.com/grblHAL/STM32F4xx/blob/daaf44d9268a259adbdc0c99d62f15abc3c8289d/Src/driver.c#L862-L916) and [synced](https://github.com/grblHAL/STM32F4xx/blob/daaf44d9268a259adbdc0c99d62f15abc3c8289d/Src/driver.c#L920-L1004) mode. Synced mode checks for new segments by comparing the segment id with the previous one so you may get som ideas from that code.

Next step could be adding an additional step interrupt handler to [stepper.c](https://github.com/grblHAL/core/blob/master/stepper.c), this to reduce overhead. It can be selected at run-time by pointing `hal.stepper.interrupt_callback` to it.

Disclaimer: I do not know if this is a workable solution or if a completely custom stepper.c is required.

**Reply by graykimchi** [reply]
Thanks for pointing me in a direction! It sounds like no one has implemented this afaik and mainly wanted to confirm.
